hal/source/gpio_device.cpp

```diff
@@ -0,0 +1,162 @@
+#include "gpio_device.h"
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+
+GPIODevice::GPIODevice(NRTFPGADevice & nrt_fpga_device) :
+	nrt_fpga_device_(nrt_fpga_device) {
+	CommonKeyPairs config = {
+	};
+	ReConfig(config);
+}
+
+
+GPIODevice::~GPIODevice() {
+	ClearInitialized();
+}
+
+error_condition GPIODevice::ReConfig(const CommonKeyPairs& config) {
+	ClearInitialized();
+	vector<uint8_t> out_data = {0, 0, 0, 0};
+	nrt_fpga_device_.Set(NRTFPGADevice::kDOData, out_data);
+	SetInitialized();
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, vector<uint8_t>& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<uint8_t> get_data;
+	result.resize(2);
+	if (key == kDIPDataName || key == kDIHDataName)
+	{
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDIData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		result[0] = get_data[0];
+		result[1] = get_data[1];
+	}
+	else if (key == kDOPDataName || key == kDOHDataName)
+	{
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDOData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		result[0] = get_data[0];
+		result[1] = get_data[1];
+	}
+	else
+		return kInvalidArgument;
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, unsigned int& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<string> destination;
+	vector<string> split_data;
+	vector<string>::iterator it;
+	boost::split(destination, key, boost::is_any_of( "." ), \
```

Comment:
IMHO, no need to use boost library here, available options are string.find() or std::regex and std::regex_match (recommended). Please also refer to https://regex101.com/ to debug the regex.